### PR TITLE
[P2075] Add Philox RNG proposal reference implementation

### PR DESCRIPTION
**Pull request by rarutyun**

### CODE REVIEW

File: rng/philox/philox.hpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        ++i;
+        if (i == n) {
+            y = philox_generate(k, x); // see below
+            increment_counter();
+            i = 0;
+        }
+        return y[i];
+    }
+
+    void discard(unsigned long long z);
```
Comment (aelizaro) [inline rng/philox/philox.hpp:107]: I implemented discard, check this out: https://godbolt.org/z/6GGbrY4G3

File: rng/philox/philox.hpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        counter_type mullo = (result << shift) >> shift;
+        return {mulhi, mullo};
+    }
+
+    static std::array<result_type, n> philox_generate(std::array<result_type, array_size> keys, std::array<counter_type, n> x)
+    {
+        for (std::size_t q = 0; q != r; ++q)
+        {
+            std::array<counter_type, n> v = x;
+            if constexpr(n == 4)
```
Comment (aelizaro) [inline rng/philox/philox.hpp:159]: we would need to add another permutations here, for n = 2, 8, 16, I can write a separate testing for all this

File: rng/philox/philox.hpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+    static consteval std::size_t ceil_log2(std::size_t val)
+    {
+        std::size_t additive = static_cast<std::size_t>(!std::has_single_bit(val));
+
+        return log2(val) + additive;
+    }
+
+    using counter_type = std::tuple_element_t<ceil_log2(w / CHAR_BIT), uint_types>;
+    using promotion_type = std::tuple_element_t<ceil_log2(w / CHAR_BIT), promotion_types>;
```
Comment (iburyl) [inline rng/philox/philox.hpp:135]: log2 is frightening...
<bit> header will help you. bit_width() is what you are looking for...

File: rng/philox/philox.hpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    static std::array<result_type, n> philox_generate(std::array<result_type, array_size> keys, std::array<counter_type, n> x)
+    {
+        for (std::size_t q = 0; q != r; ++q)
+        {
+            std::array<counter_type, n> v = x;
+            if constexpr(n == 4)
+            {
+                v[0] ^= v[2] ^= v[0] ^= v[2];
```
Comment (iburyl) [inline rng/philox/philox.hpp:161]: meh... at least leav a comment that it is an ugly swap

File: rng/philox/philox.hpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            }
+            for (std::size_t k = 0; k < array_size; ++k)
+            {
+                auto [mulhi, mullo] = mulhilo(v[2 * k], multipliers[k]);
+                x[2 * k + 1] = mullo;
+                x[2 * k] = mulhi ^ keys[k] ^ v[2 * k + 1];
+                keys[k] = (keys[k] + round_consts[k]) & result_mask;
+            }
+        }
+        return [&x]<std::size_t... Is>(std::index_sequence<Is...>) {
```
Comment (iburyl) [inline rng/philox/philox.hpp:171]: C++ developers do not look for simple solution... And when you are using hammer for too long, everything looks like a nail...

File: rng/philox/philox.hpp (hunk 3 above)
Comment (aelizaro) [reply rng/philox/philox.hpp:135]: we can probably use just a bitset<w> as a counter_type

File: rng/philox/philox.hpp (hunk 4 above)
Comment (iburyl) [reply rng/philox/philox.hpp:161]: std::swap?

**Pull request closed without merging**